app/main/views/questionnaire.py

```diff
@@ -1,38 +1,46 @@
 import logging
-from flask import render_template, request, session
-from flask_login import login_required, current_user
-from app.main.views.root import redirect_to_location
+from flask import render_template, request, session, redirect
+from flask_login import login_required
 from app.questionnaire.create_questionnaire_manager import create_questionnaire_manager
 from app.submitter.converter import SubmitterConstants
 from .. import main_blueprint
+from werkzeug.exceptions import NotFound
+from app.navigation.navigator import NavigationException
+
 
 logger = logging.getLogger(__name__)
 
 
-@main_blueprint.route('/questionnaire', methods=['GET', 'POST'])
+@main_blueprint.route('/questionnaire/<location>', methods=["GET", "POST"])
 @login_required
-def questionnaire():
-
-    logger.debug("Current user %s", current_user)
-
+def survey(location):
     # Redirect to thankyou page if the questionnaire has already been submitted
-    if SubmitterConstants.SUBMITTED_AT_KEY in session:
-        return redirect_to_location("submitted")
+    if SubmitterConstants.SUBMITTED_AT_KEY in session and location != 'thank-you':
+        return redirect('/questionnaire/thank-you')
 
     questionnaire_manager = create_questionnaire_manager()
 
-    if request.method == 'POST':
-        questionnaire_manager.process_incoming_responses(request.form)
-        current_location = questionnaire_manager.get_current_location()
-        logger.debug("POST request question - current location %s", current_location)
+    try:
+        # Go to the location in the url.
+        # This will throw an exception if invalid
+        questionnaire_manager.go_to_location(location)
+
+        # Process the POST request
+        if request.method == 'POST':
+            logger.debug("POST request question - current location %s", location)
+            questionnaire_manager.process_incoming_responses(request.form)
+            next_location = questionnaire_manager.get_current_location()
+
+            return redirect('/questionnaire/' + next_location)
 
-        return redirect_to_location(current_location)
+        context = questionnaire_manager.get_rendering_context()
+        template = questionnaire_manager.get_rendering_template()
 
-    render_data = questionnaire_manager.get_rendering_context()
+        return render_template(template,
+                               meta=context['meta'],
+                               content=context['content'],
+                               navigation=context['navigation']
+                               )
 
-    schema = questionnaire_manager.get_schema()
-    return render_template('questionnaire.html', questionnaire=render_data, data={
-        "survey_code": schema.survey_id,
-        "period_str": current_user.get_period_str(),
-        "respondent_id": current_user.get_ru_ref()
-    }, bar_title=schema.title)
+    except NavigationException:
```

Comment:
Thought the exception would be caught and the error page returned... Will update to return the error page